src/postprocess/refine.jl

```diff
@@ -0,0 +1,144 @@
+# Refinement of some quantities of interest (density, forces) following the
+# strategy described in [CDKL2021].
+#
+# [CDKL2021]:
+#     E. Cancès, G. Dusson, G. Kemlin, and A. Levitt
+#     *Practical error bounds for properties in plane-wave electronic structure
+#     calculations* Preprint, 2021. [arXiv](https://arxiv.org/abs/2111.01470)
+
+"""
+Result of calling the [`refine_scfres`](@ref) function.
+- `basis`: Refinement basis, larger than the basis used to
+           run a first [`self_consistent_field`](@ref) computation.
+- `ψ`, `ρ`, `occupation`: Quantities from the scfres, transferred to the refinement basis
+                          and with virtual orbitals removed.
+- `δψ`, `δρ`: First order corrections to the wavefunctions and density.
+              The sign is such that the refined quantities are ψ - δψ and ρ - δρ.
+"""
+struct RefinementResult
+    basis :: PlaneWaveBasis
+    ψ
+    ρ
+    occupation
+    δψ
+    δρ
+end
+
+"""
+Transfer the result of an SCF to a larger basis set,
+and compute first order corrections ("refinements") to the wavefunctions and density.
+
+Returns a [`RefinementResult`](@ref) instance that can be used to refine quantities of interest,
+through [`refine_density`](@ref) and [`refine_forces`](@ref).
+"""
+function refine_scfres(scfres, basis_ref::PlaneWaveBasis{T}; ΩpK_tol,
+                       occ_threshold=default_occupation_threshold(T), kwargs...) where {T}
+    basis = scfres.basis
+
+    @assert basis.model.lattice == basis_ref.model.lattice
+    @assert length(basis.kpoints) == length(basis_ref.kpoints)
+    @assert all(basis.kpoints[ik].coordinate == basis_ref.kpoints[ik].coordinate
+                for ik in 1:length(basis.kpoints))
```

Comment:
This needs an assertion about occupation / temperature somewhere (maybe taken care of elsewhere), just to make sure that for metals this fails.